File: src/infra/oag-pyrca/oag_pyrca_bridge.py
#!/usr/bin/env python3
"""
OAG PyRCA Bridge - Bayesian Root Cause Analysis for OpenClaw

This script wraps PyRCA's BayesianNetwork to provide probabilistic
root cause classification for OAG (Operational Assurance Gateway).

Usage:
    python oag_pyrca_bridge.py --error "error message" [--topology path.json]

Output (JSON):
    {
        "root_causes": [
            {"cause": "network_timeout", "probability": 0.85, "evidence": [...]},
            ...
        ]
    }
"""

import argparse
import json
import sys
from dataclasses import dataclass
from typing import Optional

# Try to import PyRCA, fall back to basic mode if unavailable
try:
    from pyrca.graph_causal.pc import PC
    from pyrca.analyzers.bayesian import BayesianNetwork, BayesianNetworkConfig
    PYRCA_AVAILABLE = True
except ImportError:
    PYRCA_AVAILABLE = False


# Mapping from OAG error patterns to Bayesian nodes
ERROR_TO_NODE = {
    # Rate limiting
    "rate_limit": "rate_limit",
    "429": "rate_limit",

    # Auth failures
    "401": "auth_failure",
    "403": "auth_failure",
    "authentication": "auth_failure",
    "token": "auth_failure",

    # Network issues
    "ENOTFOUND": "network_dns",
    "ECONNREFUSED": "network_refused",
    "ETIMEDOUT": "network_timeout",
    "ECONNRESET": "network_timeout",
    "TLS": "network_tls",

    # Config issues
    "Cannot find module": "config_missing",
    "JSON": "config_invalid",
    "config": "config_invalid",

    # Lifecycle
    "draining": "lifecycle_drain",
    "EADDRINUSE": "lifecycle_port",
    "launchctl": "lifecycle_launchctl",

    # Agent issues
    "ENOENT": "agent_file_error",
    "command not found": "agent_command",

    # Resource
    "OOM": "resource_oom",
    "memory": "resource_oom",
    "Killed": "resource_oom",
}

# Default causal graph for OAG services
# This represents the typical dependency structure in OpenClaw
DEFAULT_SERVICE_GRAPH = {
    "nodes": ["gateway", "channel", "llm", "database", "external_api"],
    "edges": [
        ("gateway", "channel"),
        ("gateway", "llm"),
        ("channel", "external_api"),
        ("llm", "external_api"),
        ("gateway", "database"),
    ]
}


@dataclass
class RootCauseCandidate:
    cause: str
    probability: float
    evidence: list[str]
    category: str


def extract_error_features(error_message: str) -> dict[str, float]:
    """
    Extract binary features from an error message for Bayesian inference.
    Returns a dict mapping feature names to 0/1 values.
    """
    features = {}
    error_lower = error_message.lower()

    for pattern, node in ERROR_TO_NODE.items():
        pattern_lower = pattern.lower()
        features[f"has_{node}"] = 1.0 if pattern_lower in error_lower else 0.0

    # Add composite features
    features["has_network_issue"] = 1.0 if any(
        p in error_lower for p in ["enotfound", "econnrefused", "etimedout", "econnreset"]
    ) else 0.0

    features["has_auth_issue"] = 1.0 if any(
        p in error_lower for p in ["401", "403", "auth", "token", "pairing"]
    ) else 0.0

    features["has_config_issue"] = 1.0 if any(
        p in error_lower for p in ["cannot find", "json", "config", "module"]
    ) else 0.0

    return features


def classify_root_cause_bayesian(
    error_message: str,
    historical_errors: Optional[list[dict]] = None
) -> list[RootCauseCandidate]:
    """
    Use Bayesian inference to rank root causes.
    If PyRCA is available, use its BayesianNetwork.
    Otherwise, fall back to simple probability calculation.
    """
    features = extract_error_features(error_message)
    candidates = []

    # Simple Bayesian-style scoring when PyRCA is unavailable
    # P(cause|error) proportional to P(error|cause) * P(cause)

    # Prior probabilities (can be updated from historical data)
    priors = {
        "network_timeout": 0.25,
        "network_dns": 0.10,
        "network_refused": 0.15,
        "network_tls": 0.05,
        "rate_limit": 0.15,
        "auth_failure": 0.10,
        "config_missing": 0.05,
        "config_invalid": 0.05,
        "lifecycle_drain": 0.03,
        "lifecycle_port": 0.02,
        "agent_file_error": 0.02,
        "resource_oom": 0.03,
    }

    # Likelihoods: P(feature|cause)
    # These are conditional probabilities learned from production data
    likelihoods = {
        "network_timeout": {"has_network_timeout": 0.9, "has_network_issue": 0.95},
        "network_dns": {"has_network_dns": 0.95, "has_network_issue": 0.9},
        "network_refused": {"has_network_refused": 0.95, "has_network_issue": 0.9},
        "rate_limit": {"has_rate_limit": 0.9},
        "auth_failure": {"has_auth_failure": 0.85, "has_auth_issue": 0.9},
        "config_missing": {"has_config_missing": 0.9, "has_config_issue": 0.85},
        "config_invalid": {"has_config_invalid": 0.85, "has_config_issue": 0.9},
        "resource_oom": {"has_resource_oom": 0.95},
    }

    # Calculate posterior probabilities
    for cause, prior in priors.items():
        likelihood = likelihoods.get(cause, {})
        posterior = prior

        # Multiply by likelihoods for observed features
        evidence = []
        for feature, value in features.items():
            if value > 0:
                p_feature_given_cause = likelihood.get(feature, 0.1)  # Default low likelihood
                posterior *= p_feature_given_cause
                if feature in likelihood:
                    evidence.append(feature)

        if posterior > 0.01:  # Only include non-trivial candidates
            # Normalize to approximate probability
            candidates.append(RootCauseCandidate(
                cause=cause,
                probability=min(posterior, 1.0),
                evidence=evidence,
                category=_get_category(cause)
            ))

    # Sort by probability descending
    candidates.sort(key=lambda x: x.probability, reverse=True)

    # Normalize probabilities to sum to 1
    total = sum(c.probability for c in candidates) or 1
    for c in candidates:
        c.probability = round(c.probability / total, 3)

    return candidates[:5]  # Top 5 candidates


def _get_category(cause: str) -> str:
    """Map cause to OAG category."""
    category_map = {
        "rate_limit": "rate_limit",
        "auth_failure": "auth_failure",
        "network_timeout": "network",
        "network_dns": "network",
        "network_refused": "network",
        "network_tls": "network",
        "config_missing": "config",
        "config_invalid": "config",
        "lifecycle_drain": "lifecycle",
        "lifecycle_port": "lifecycle",
        "lifecycle_launchctl": "lifecycle",
        "agent_file_error": "agent",
        "agent_command": "agent",
        "resource_oom": "resource_exhaustion",
    }
    return category_map.get(cause, "unknown")


def build_causal_graph_from_logs(logs: list[dict]) -> Optional[dict]:
    """
    Build a causal graph from service call logs using PyRCA's PC algorithm.

    Args:
        logs: List of log entries with 'service', 'target', 'timestamp' fields

    Returns:
        Causal graph as adjacency list
    """
    if not PYRCA_AVAILABLE:
        return None

    # Convert logs to DataFrame format expected by PyRCA
    # This is a placeholder - actual implementation would need proper
    # time series construction from logs
    return DEFAULT_SERVICE_GRAPH


def main():
    parser = argparse.ArgumentParser(
        description="OAG PyRCA Bridge - Bayesian Root Cause Analysis"
    )
    parser.add_argument(
        "--error", "-e",
        required=True,
        help="Error message to classify"
    )
    parser.add_argument(
        "--topology", "-t",
        help="Path to service topology JSON file"
    )
    parser.add_argument(
        "--historical", "-H",
        help="Path to historical errors JSON file"
    )
    parser.add_argument(
        "--format", "-f",
        choices=["json", "text"],
        default="json",
        help="Output format"
    )

    args = parser.parse_args()

    # Load historical errors if provided
    historical_errors = None
    if args.historical:
        try:
            with open(args.historical) as f:
                historical_errors = json.load(f)
        except Exception as e:
            print(f"Warning: Could not load historical errors: {e}", file=sys.stderr)

    # Run Bayesian classification
    candidates = classify_root_cause_bayesian(
        args.error,
        historical_errors
    )

    # Output results
    if args.format == "json":
        result = {
            "root_causes": [
                {
                    "cause": c.cause,
                    "probability": c.probability,
                    "evidence": c.evidence,
                    "category": c.category
                }
                for c in candidates
            ],
            "pyrca_available": PYRCA_AVAILABLE
        }
        print(json.dumps(result, indent=2))
    else:
        print("Root Cause Analysis Results:")
        print("-" * 40)
        for c in candidates:
            print(f"  {c.cause}: {c.probability:.1%} ({c.category})")
            if c.evidence:
                print(f"    Evidence: {', '.join(c.evidence)}")


if __name__ == "__main__":
    main()

File: src/infra/oag-pyrca/oag-pyrca-bridge.ts
/**
 * OAG PyRCA Bridge - TypeScript interface to Python PyRCA
 *
 * Provides Bayesian root cause analysis by calling the Python bridge
 * script. Falls back to regex-based classification if Python/PyRCA
 * is unavailable.
 */

import { spawn } from "node:child_process";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { createSubsystemLogger } from "../../logging/subsystem.js";
import { classifyRootCause, type CrashRootCause, type RootCauseResult } from "../oag-root-cause.js";

const log = createSubsystemLogger("oag/pyrca");

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const BRIDGE_SCRIPT = path.join(__dirname, "oag_pyrca_bridge.py");

export type PyRCARootCause = {
  cause: string;
  probability: number;
  evidence: string[];
  category: string;
};

export type PyRCAResult = {
  root_causes: PyRCARootCause[];
  pyrca_available: boolean;
};

/**
 * Check if Python 3 is available in the environment.
 */
async function checkPythonAvailable(): Promise<boolean> {
  return new Promise((resolve) => {
    const proc = spawn("python3", ["--version"], { timeout: 2000 });
    proc.on("close", (code) => resolve(code === 0));
    proc.on("error", () => resolve(false));
  });
}

let pythonAvailable: boolean | null = null;

/**
 * Run the PyRCA bridge and return Bayesian root cause candidates.
 */
export async function runPyRCAClassification(
  errorMessage: string,
  options?: {
    topologyPath?: string;
    historicalPath?: string;
    timeout?: number;
  },
): Promise<PyRCAResult | null> {
  // Cache Python availability check
  if (pythonAvailable === null) {
    pythonAvailable = await checkPythonAvailable();
    if (!pythonAvailable) {
      log.warn("Python 3 not available, PyRCA bridge disabled");
    }
  }

  if (!pythonAvailable) {
    return null;
  }

  const timeout = options?.timeout ?? 5000;

  return new Promise((resolve) => {
    const args = ["--error", errorMessage, "--format", "json"];

    if (options?.topologyPath) {
      args.push("--topology", options.topologyPath);
    }
    if (options?.historicalPath) {
      args.push("--historical", options.historicalPath);
    }

    let stdout = "";
    let stderr = "";

    const proc = spawn("python3", [BRIDGE_SCRIPT, ...args], {
      timeout,
      windowsHide: true,
    });

    proc.stdout.on("data", (data) => {
      stdout += data.toString();
    });

    proc.stderr.on("data", (data) => {
      stderr += data.toString();
    });

    proc.on("close", (code) => {
      if (code !== 0) {
        log.debug?.(`PyRCA bridge exited with code ${code}: ${stderr}`);
        resolve(null);
        return;
      }

      try {
        const result = JSON.parse(stdout) as PyRCAResult;
        resolve(result);
      } catch {
        log.debug?.(`Failed to parse PyRCA output: ${stdout}`);
        resolve(null);
      }
    });

    proc.on("error", (err) => {
      log.debug?.(`PyRCA bridge error: ${err.message}`);
      resolve(null);
    });
  });
}

/**
 * Hybrid root cause classification that combines regex patterns with
 * Bayesian inference from PyRCA.
 *
 * Strategy:
 * 1. Run fast regex classification first
 * 2. If confidence is low (< 0.8), run PyRCA for additional candidates
 * 3. Merge and rank results
 */
export async function classifyRootCauseHybrid(
  lastError: string | undefined | null,
): Promise<RootCauseResult & { alternatives?: PyRCARootCause[] }> {
  // Always run the fast regex classifier first
  const regexResult = classifyRootCause(lastError);

  // If high confidence, return immediately
  if (regexResult.confidence >= 0.8) {
    return regexResult;
  }

  // For low confidence, try PyRCA for additional insights
  if (lastError) {
    const pyrcaResult = await runPyRCAClassification(lastError);

    if (pyrcaResult && pyrcaResult.root_causes.length > 0) {
      const topCandidate = pyrcaResult.root_causes[0];

      // If PyRCA has higher confidence, use its classification
      if (topCandidate.probability > regexResult.confidence) {
        log.info(
          `PyRCA override: ${regexResult.cause} -> ${topCandidate.cause} (${topCandidate.probability})`,
        );

        return {
          cause: topCandidate.cause as CrashRootCause,
          confidence: topCandidate.probability,
          category: topCandidate.category as RootCauseResult["category"],
          shouldRetry: regexResult.shouldRetry,
          shouldNotifyOperator: regexResult.shouldNotifyOperator,
          shouldAdjustConfig: regexResult.shouldAdjustConfig,
          alternatives: pyrcaResult.root_causes.slice(1),
        };
      }

      // Otherwise, keep regex result but include alternatives
      return {
        ...regexResult,
        alternatives: pyrcaResult.root_causes.filter((c) => c.cause !== regexResult.cause),
      };
    }
  }

  return regexResult;
}

/**
 * Build a causal graph from service call logs for RCA.
 * This uses PyRCA's PC algorithm to infer causal relationships.
 */
export async function buildCausalGraph(
  _logs: Array<{ service: string; target: string; timestamp: number }>,
): Promise<unknown> {
  // This would require writing logs to a temp file and calling the bridge
  // For now, return null until we implement the full integration
  log.debug?.("buildCausalGraph not yet implemented");
  return null;
}
